src/test/java/com/google/cloud/spanner/pgadapter/python/PythonTransactionTests.java

```diff
@@ -0,0 +1,1095 @@
+    Assert.assertEquals(3, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+
+    List<String> sql = new ArrayList<>();
+
+    for (ExecuteSqlRequest request : mockSpanner.getRequestsOfType(ExecuteSqlRequest.class)) {
+      sql.add(request.getSql());
+    }
+
+    Assert.assertEquals(1, Collections.frequency(sql, sql1));
+    Assert.assertEquals(1, Collections.frequency(sql, sql2));
+    Assert.assertEquals(0, Collections.frequency(sql, sql3));
+    Assert.assertEquals(1, Collections.frequency(sql, sql4));
+
+    mockSpanner.clearRequests();
+    statements.set(9, "set session isolation_level 4");
+
+    actualOutput = executeTransactions(pgServer.getLocalPort(), statements);
+    Assert.assertEquals(expectedOutput, actualOutput);
+
+    for (AbstractMessage request : mockSpanner.getRequests())
+      System.out.println(request.getClass());
+
+    Assert.assertEquals(1, mockSpanner.countRequestsOfType(CommitRequest.class));
+    Assert.assertEquals(0, mockSpanner.countRequestsOfType(RollbackRequest.class));
+    Assert.assertEquals(3, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+
+    sql.clear();
+
+    for (ExecuteSqlRequest request : mockSpanner.getRequestsOfType(ExecuteSqlRequest.class)) {
+      sql.add(request.getSql());
+    }
+
+    Assert.assertEquals(1, Collections.frequency(sql, sql1));
+    Assert.assertEquals(1, Collections.frequency(sql, sql2));
+    Assert.assertEquals(0, Collections.frequency(sql, sql3));
+    Assert.assertEquals(1, Collections.frequency(sql, sql4));
+
+    Assert.assertEquals(
+        2,
+        getWireMessagesOfType(QueryMessage.class).stream()
+            .filter(qm -> qm.toString().contains("READ COMMITTED"))
+            .count());
+  }
+
+  private void restartServerWithDifferentVersion(String version) throws Exception {
+    stopMockSpannerAndPgAdapterServers();
+    doStartMockSpannerAndPgAdapterServers("d", Arrays.asList("-v", version));
+    Assert.assertEquals(version, pgServer.getOptions().getServerVersion());
+  }
+
+  @Test
+  public void testIsolationLevelInTransactions() throws IOException, InterruptedException {
+    // tests isolation_level settings using connection.isolation_level variable
+    List<String> statements = new ArrayList<>();
+    String sql1 = "Select * from some_table";
+    String sql2 = "insert into some_table(col1, col2) values(value1, value2)";
+    String sql3 = "Select * from some_table3";
+    String sql4 = "insert into some_table3(col1, col2) values(value1, value2)";
+
+    statements.add("query");
+    statements.add(sql1);
+
+    statements.add("update");
+    statements.add(sql2);
+
+    statements.add("update");
+    statements.add(sql4);
+
+    statements.add("transaction");
+    statements.add("commit");
+
+    // This will not cause error because the server version is 1.0,
+    // so the psycopg2 will send SERIALIZABLE instead of REPEATABLE READ as the isolation level
```

Comment:
It would be good to add a more elaborate comment on this. For someone who is not familiar with PostgreSQL / psycopg2, this comment does not explain why and when there is a difference.

Reply:
Done